{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "from Bio import SeqIO\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "def extract_isoform_of(description: str) -> str:\n",
    "    \"\"\"\n",
    "    Extracts protein of origin from the description.\n",
    "\n",
    "    :param description: The description part of the fasta entry.\n",
    "    :type description: str\n",
    "    :return: Native protein.\n",
    "    :rtype: str\n",
    "    \"\"\"\n",
    "    isoform_parts = description.split(\"Isoform of\")\n",
    "    if len(isoform_parts) > 1:\n",
    "        isoform_info = isoform_parts[1].strip().split(\",\")[0]\n",
    "        return isoform_info.strip()\n",
    "    return None\n",
    "\n",
    "def extract_functional_traits(description: str) -> str:\n",
    "    \"\"\"\n",
    "    Extracts functional traits from the description.\n",
    "\n",
    "    :param description: The description part of the fasta entry.\n",
    "    :type description: str\n",
    "    :return: Functional traits.\n",
    "    :rtype: str\n",
    "    \"\"\"\n",
    "    comma_index = description.find(\",\")\n",
    "    os_index = description.find(\"OS=\")\n",
    "    if comma_index != -1 and os_index != -1:\n",
    "        return description[comma_index + 1:os_index].strip()\n",
    "    else:\n",
    "        return None\n",
    "\n",
    "\n",
    "def extract_organism_name(description: str) -> str:\n",
    "    \"\"\"\n",
    "    Extracts organism name from the description.\n",
    "\n",
    "    :param description: The description part of the fasta entry.\n",
    "    :type description: str\n",
    "    :return: Organism name.\n",
    "    :rtype: str\n",
    "    \"\"\"\n",
    "    organism_start = description.find(\"OS=\")\n",
    "    if organism_start != -1:\n",
    "        organism_end = description.find(\" \", organism_start + 3)\n",
    "        if organism_end != -1:\n",
    "            organism_end = description.find(\" \", organism_end + 1)\n",
    "            if organism_end == -1:\n",
    "                organism_end = len(description)\n",
    "            return description[organism_start + 3:organism_end].strip()\n",
    "    return None\n",
    "\n",
    "\n",
    "def extract_gene_name(description: str) -> str:\n",
    "    \"\"\"\n",
    "    Extracts gene name from the description.\n",
    "\n",
    "    :param description: The description part of the fasta entry.\n",
    "    :type description: str\n",
    "    :return: Gene name.\n",
    "    :rtype: str\n",
    "    \"\"\"\n",
    "    gene_start = description.find(\"GN=\")\n",
    "    if gene_start != -1:\n",
    "        gene_end = description.find(\" \", gene_start)\n",
    "        if gene_end == -1:\n",
    "            gene_end = len(description)\n",
    "        return description[gene_start + 3:gene_end].strip()\n",
    "    else:\n",
    "        return None\n",
    "\n",
    "def fasta_to_df(file_path: str) -> pd.DataFrame:\n",
    "    \"\"\"\n",
    "    Convert a FASTA file to a pandas DataFrame.\n",
    "\n",
    "    :param file_path: Path to the FASTA file.\n",
    "    :type file_path: str\n",
    "    :return: DataFrame containing FASTA information.\n",
    "    :rtype: pd.DataFrame\n",
    "    \"\"\"\n",
    "\n",
    "    data = []\n",
    "    for record in SeqIO.parse(file_path, \"fasta\"):\n",
    "        header_info = record.id.split(\"|\")\n",
    "        description = record.description\n",
    "        data.append({\n",
    "            \"db\": header_info[0],\n",
    "            \"unique_identifier\": header_info[1],\n",
    "            \"entry_name\": header_info[2],\n",
    "            \"isoform_of\": extract_isoform_of(description),\n",
    "            \"functional_traits\": extract_functional_traits(description),\n",
    "            \"organism_name\": extract_organism_name(description),\n",
    "            \"gene_name\": extract_gene_name(description),\n",
    "            \"protein_sequence\": str(record.seq)\n",
    "        })\n",
    "\n",
    "    fasta = pd.DataFrame(data)\n",
    "    return fasta\n",
    "\n",
    "\n",
    "class IsoRetriever:\n",
    "    def __init__(self, file_path):\n",
    "        self.file_path = file_path\n",
    "        self._read_csv()\n",
    "\n",
    "    def _read_csv(self):\n",
    "        data_df = pd.read_csv(self.file_path)\n",
    "        \n",
    "        if \"UniprotA\" in data_df.columns and \"UniprotB\" in data_df.columns:\n",
    "            self._read_csv_as_v2(data_df)\n",
    "        else:\n",
    "            self._read_csv_as_v1(data_df)\n",
    "\n",
    "    def _read_csv_as_v1(self, data_df):\n",
    "        uniprotA_293t = data_df[[\"UniprotA-293T\", \"SymbolA-293T\"]].dropna().rename(columns={\"UniprotA-293T\": \"isoform\", \"SymbolA-293T\": \"gene_name\"})\n",
    "        uniprotB_293t = data_df[[\"UniprotB-293T\", \"SymbolB-293T\"]].dropna().rename(columns={\"UniprotB-293T\": \"isoform\", \"SymbolB-293T\": \"gene_name\"})\n",
    "        uniprotA_hct116 = data_df[[\"UniprotA-HCT116\", \"SymbolA-HCT116\"]].dropna().rename(columns={\"UniprotA-HCT116\": \"isoform\", \"SymbolA-HCT116\": \"gene_name\"})\n",
    "        uniprotB_hct116 = data_df[[\"UniprotB-HCT116\", \"SymbolB-HCT116\"]].dropna().rename(columns={\"UniprotB-HCT116\": \"isoform\", \"SymbolB-HCT116\": \"gene_name\"})        \n",
    "        merged_df = pd.concat([uniprotA_293t, uniprotB_293t, uniprotA_hct116, uniprotB_hct116], ignore_index=True)\n",
    "        self.unique_isoforms_df = merged_df.drop_duplicates(subset=[\"isoform\"]).dropna()\n",
    "\n",
    "    def _read_csv_as_v2(self, data_df):\n",
    "        uniprotA_df = data_df[[\"UniprotA\", \"SymbolA\"]].dropna().rename(columns={\"UniprotA\": \"isoform\", \"SymbolA\": \"gene_name\"})\n",
    "        uniprotB_df = data_df[[\"UniprotB\", \"SymbolB\"]].dropna().rename(columns={\"UniprotB\": \"isoform\", \"SymbolB\": \"gene_name\"})\n",
    "        merged_df = pd.concat([uniprotA_df, uniprotB_df], ignore_index=True)\n",
    "        self.unique_isoforms_df = merged_df.drop_duplicates(subset=[\"isoform\"]).dropna()\n",
    "\n",
    "    def check_duplicates_iso_retriever(self): \n",
    "        duplicates_exist = self.unique_isoforms_df.duplicated(subset=[\"isoform\"]).any()\n",
    "        if duplicates_exist:\n",
    "            print(\"DETECTED duplicates in 'isoform'.\")\n",
    "        else:\n",
    "            print(\"There're NO duplicated values in 'isoform'.\")\n",
    "        return duplicates_exist\n",
    "\n",
    "    def get_unique_isoforms_df(self):\n",
    "        return self.unique_isoforms_df\n",
    "\n",
    "\n",
    "class iso_lib_matcher:\n",
    "    def __init__(self, lib_df, dataset_df):\n",
    "        self.lib_df = lib_df\n",
    "        self.dataset_df = dataset_df\n",
    "        self.filter()\n",
    "\n",
    "    def filter(self):\n",
    "        lib_column_2 = \"unique_identifier\"\n",
    "        lib_column_4 = \"isoform_of\"\n",
    "        lib_column_7 = \"gene_name\"\n",
    "        dataset_column = \"isoform\"\n",
    "        self.filtered_df = self.lib_df[self.lib_df[lib_column_2].isin(self.dataset_df[dataset_column].dropna())][[lib_column_4, lib_column_2, lib_column_7]]\n",
    "        self.filtered_df.columns = [\"canonical_protein\", \"isoform\", lib_column_7]\n",
    "\n",
    "    def check_duplicates_iso_lib_match(self): \n",
    "        duplicates_exist = self.filtered_df.duplicated(subset=[\"isoform\"]).any()\n",
    "        if duplicates_exist:\n",
    "            print(\"DETECTED duplicates in 'isoform'.\")\n",
    "        else:\n",
    "            print(\"There're NO duplicates in 'isoform'.\")\n",
    "        return duplicates_exist\n",
    "\n",
    "    def get_filtered_df(self):\n",
    "        return self.filtered_df\n",
    "\n",
    "class can_lib_matcher:\n",
    "    def __init__(self, dataset_df, lib_df):\n",
    "        self.dataset_df = dataset_df\n",
    "        self.lib_df = lib_df\n",
    "        self.filter()\n",
    "\n",
    "    def filter(self):\n",
    "        dataset_column_1 = \"isoform\"\n",
    "        dataset_column_2 = \"gene_name_x\"\n",
    "        lib_column_4 = \"isoform_of\"\n",
    "        merged_df = pd.merge(self.dataset_df, self.lib_df, how=\"inner\", left_on=dataset_column_1, right_on=lib_column_4)\n",
    "        self.filtered_df = merged_df[[lib_column_4, dataset_column_1, dataset_column_2]].drop_duplicates(subset=[dataset_column_1])\n",
    "        self.filtered_df.columns = [\"canonical_protein\", dataset_column_1, \"gene_name\"]\n",
    "\n",
    "    def check_duplicates_can_lib_match(self): \n",
    "        duplicates_exist = self.filtered_df.duplicated(subset=[\"isoform\"]).any()\n",
    "        if duplicates_exist:\n",
    "            print(\"DETECTED duplicates in 'isoform'.\")\n",
    "        else:\n",
    "            print(\"There're NO duplicates in 'isoform'.\")\n",
    "        return duplicates_exist\n",
    "\n",
    "    def get_filtered_df(self):\n",
    "        return self.filtered_df\n",
    "\n",
    "class gene_lib_matcher:\n",
    "    def __init__(self, dataset_df, lib_df):\n",
    "        self.dataset_df = dataset_df\n",
    "        self.lib_df = lib_df\n",
    "        self.filter()\n",
    "\n",
    "    def filter(self):\n",
    "        dataset_column_1 = \"isoform\"\n",
    "        dataset_column_2 = \"gene_name\"\n",
    "        lib_column_4 = \"isoform_of\"\n",
    "        lib_column_7 = \"gene_name\"\n",
    "        merged_df = pd.merge(self.dataset_df, self.lib_df, how=\"inner\", left_on=dataset_column_2, right_on=lib_column_7)\n",
    "        self.filtered_df = merged_df[[lib_column_4, dataset_column_1, dataset_column_2]].drop_duplicates(subset=[dataset_column_1])\n",
    "        self.filtered_df.columns = [\"canonical_protein\", dataset_column_1, dataset_column_2]\n",
    "\n",
    "    def check_duplicates_gene_lib_match(self): \n",
    "        duplicates_exist = self.filtered_df.duplicated(subset=[\"isoform\"]).any()\n",
    "        if duplicates_exist:\n",
    "            print(\"DETECTED duplicates in 'isoform'.\")\n",
    "        else:\n",
    "            print(\"There're NO duplicates in 'isoform'.\")\n",
    "        return duplicates_exist\n",
    "\n",
    "    def get_filtered_df(self):\n",
    "        return self.filtered_df\n",
    "    \n",
    "class addition:\n",
    "    def __init__(self, dataset_1, dataset_2, dataset_3, column_drop_duplicates):\n",
    "        self.dataset_1 = dataset_1\n",
    "        self.dataset_2 = dataset_2\n",
    "        self.dataset_3 = dataset_3\n",
    "        self.column_drop_duplicates = column_drop_duplicates\n",
    "        self.merge()\n",
    "\n",
    "    def merge(self):\n",
    "        merged_df = pd.concat([self.dataset_1, self.dataset_2, self.dataset_3], ignore_index=True)\n",
    "        self.merged_df = merged_df.drop_duplicates(subset=[self.column_drop_duplicates])\n",
    "\n",
    "    def check_duplicates_lib_matches(self): \n",
    "        duplicates_exist = self.merged_df.duplicated(subset=[\"isoform\"]).any()\n",
    "        if duplicates_exist:\n",
    "            print(\"DETECTED duplicates in 'isoform'.\")\n",
    "        else:\n",
    "            print(\"There're NO duplicates in 'isoform'.\")\n",
    "        return duplicates_exist\n",
    "    \n",
    "    def get_merged_df(self):\n",
    "        return self.merged_df\n",
    "    \n",
    "\n",
    "class subtract:\n",
    "    def __init__(self, add_df, subtract_df, add_column, subtract_column):\n",
    "        self.add_df = add_df\n",
    "        self.subtract_df = subtract_df\n",
    "        self.add_column = add_column\n",
    "        self.subtract_column = subtract_column\n",
    "\n",
    "    def subtraction(self):\n",
    "        sum = self.add_df[~self.add_df[self.add_column].isin(self.subtract_df[self.subtract_column])]\n",
    "        return sum\n",
    "    \n",
    "class SortUndefined:\n",
    "    def __init__(self, dataset):\n",
    "        self.dataset = dataset\n",
    "\n",
    "    def sort(self):\n",
    "        data_column_1 = \"isoform\"\n",
    "        data_column_2 = \"gene_name\"\n",
    "        modified_df = self.dataset.copy()\n",
    "        modified_df['new_column'] = modified_df[data_column_2]\n",
    "        modified_df = modified_df.loc[:, ['new_column', data_column_1, data_column_2]]\n",
    "        modified_df.columns = [\"canonical_protein\", data_column_1, data_column_2]\n",
    "        \n",
    "        return modified_df\n",
    "    \n",
    "class IsoformCount:\n",
    "    def __init__(self, dataset):\n",
    "        self.dataset = dataset\n",
    "        self.iso_count()\n",
    "\n",
    "    def iso_count(self):\n",
    "        isoform_dict = {}\n",
    "\n",
    "        for index, row in self.dataset.iterrows():\n",
    "            data_column_1 = row['canonical_protein']\n",
    "            data_column_2 = row['isoform']\n",
    "            if data_column_1 in isoform_dict:\n",
    "                isoform_dict[data_column_1].append(data_column_2)\n",
    "            else:\n",
    "                isoform_dict[data_column_1] = [data_column_2]\n",
    "\n",
    "        for data_column_1, data_column_2_values in isoform_dict.items():\n",
    "            for i, data_column_2_value in enumerate(data_column_2_values, start=1):\n",
    "                self.dataset.loc[self.dataset['canonical_protein'] == data_column_1, f'isoform_{i}'] = data_column_2_value\n",
    "\n",
    "        self.dataset.drop_duplicates(subset=['canonical_protein'], inplace=True)\n",
    "        dataset_count = self.dataset[['canonical_protein'] + [col for col in self.dataset.columns if col.startswith('isoform_')]].copy()\n",
    "        dataset_count.reset_index(drop=True, inplace=True)\n",
    "        dataset_count['isoform_count'] = dataset_count.filter(like='isoform_').notna().sum(axis=1)\n",
    "        self.dataset_count = dataset_count\n",
    "\n",
    "    def print_number_of_isoforms(self):\n",
    "        total_isoform_count = self.dataset_count['isoform_count'].sum()\n",
    "        print(\"Total Isoform Count:\", total_isoform_count)\n",
    "\n",
    "    def get_dataset_count(self):\n",
    "        return self.dataset_count\n",
    "    \n",
    "class IsoformDistribution:\n",
    "    def __init__(self, dataframe):\n",
    "        self.dataframe = dataframe\n",
    "\n",
    "    def iso_dis(self):\n",
    "        count_df = self.dataframe['isoform_count'].value_counts().reset_index()\n",
    "        count_df.columns = ['isoform_count', '# of entries']\n",
    "        count_df = count_df.sort_values(by='isoform_count')\n",
    "        total_entries = count_df['# of entries'].sum()\n",
    "        count_df['frequency (%)'] = count_df['# of entries'] / total_entries * 100\n",
    "\n",
    "        return count_df, total_entries\n",
    "\n",
    "    def plot_distribution(self):\n",
    "        count_df, _ = self.iso_dis()\n",
    "        fig, ax1 = plt.subplots(figsize=(10, 6))\n",
    "\n",
    "        ax1.bar(count_df['isoform_count'], count_df['frequency (%)'], color='skyblue')\n",
    "        ax1.set_title('Isoform Distribution')\n",
    "        ax1.set_xlabel('Number of Isoforms')\n",
    "        ax1.set_ylabel('Frequency (%)')\n",
    "        ax1.grid(True)\n",
    "        ax1.set_xticks(range(int(count_df['isoform_count'].max()) + 1))\n",
    "        ax1.tick_params(axis='both', which='major')\n",
    "\n",
    "        zoomed_data = count_df[count_df['frequency (%)'] < 1]\n",
    "        ax2 = ax1.inset_axes([0.6, 0.6, 0.3, 0.3])  \n",
    "        ax2.bar(zoomed_data['isoform_count'], zoomed_data['frequency (%)'], color='skyblue')\n",
    "        ax2.set_xlabel('Number of Isoforms')\n",
    "        ax2.set_ylabel('Frequency (%)')\n",
    "        ax2.set_xticks(zoomed_data['isoform_count'])\n",
    "        ax2.tick_params(axis='both', which='major')\n",
    "        ax2.set_ylim(0, 1.2 * zoomed_data['frequency (%)'].max())\n",
    "\n",
    "        plt.tight_layout()\n",
    "        plt.show()    \n",
    "\n",
    "\n",
    "class protein_matcher:\n",
    "    def __init__(self, dataset, protein_list, dataset_column, protein_list_column):\n",
    "        self.dataset = dataset\n",
    "        self.protein_list = protein_list\n",
    "        self.dataset_column = dataset_column\n",
    "        self.protein_list_column = protein_list_column\n",
    "        self.filter()\n",
    "\n",
    "    def filter(self):\n",
    "        self.filtered_df = self.dataset[self.dataset[self.dataset_column].isin(self.protein_list[self.protein_list_column].dropna())]\n",
    "\n",
    "    def check_duplicates_protein_match(self): \n",
    "        duplicates_exist = self.filtered_df.duplicated(subset=[\"canonical_protein\"]).any()\n",
    "        if duplicates_exist:\n",
    "            print(\"DETECTED duplicates in 'canonical_protein'.\")\n",
    "        else:\n",
    "            print(\"There're NO duplicates in 'canonical_protein'.\")\n",
    "        return duplicates_exist\n",
    "\n",
    "    def get_filtered_df(self):\n",
    "        return self.filtered_df\n",
    "\n",
    "class IsoformExtractor:\n",
    "    def __init__(self, dataframe):\n",
    "        self.dataframe = dataframe\n",
    "        self.extract_isoforms()\n",
    "\n",
    "    def extract_isoforms(self):\n",
    "        isoform_columns = [col for col in self.dataframe.columns if col.startswith(\"isoform_\")]\n",
    "        isoform_values = []\n",
    "        for col in isoform_columns:\n",
    "            isoform_values.extend(self.dataframe[col].dropna().unique())\n",
    "        isoform_df = pd.DataFrame({\"isoform\": isoform_values})\n",
    "        isoform_df = isoform_df[~isoform_df['isoform'].astype(str).str.isdigit()]\n",
    "        self.unique_isoforms_df = isoform_df.drop_duplicates(subset=[\"isoform\"]).reset_index(drop=True)\n",
    "\n",
    "    def get_unique_isoforms_df(self):\n",
    "        return self.unique_isoforms_df\n",
    "    \n",
    "class UniqueInteractionsHandler:\n",
    "    def __init__(self, dataframe):\n",
    "        self.dataframe = dataframe.dropna(subset=['UniprotA', 'UniprotB'])\n",
    "\n",
    "    def process_ppi(self, protein_match_df):\n",
    "        match_list = []\n",
    "        unique_isoforms = set(protein_match_df['isoform'].dropna())\n",
    "        for index, row in self.dataframe.iterrows():\n",
    "            if row['UniprotA'] in unique_isoforms and row['UniprotB'] in unique_isoforms:\n",
    "                match_list.append(row)\n",
    "\n",
    "        self.dataframe = pd.DataFrame(match_list)\n",
    "        unique_uniprot_values = pd.concat([self.dataframe['UniprotA'], self.dataframe['UniprotB']]).dropna().unique()\n",
    "        return pd.DataFrame({'isoform': unique_uniprot_values})\n",
    "\n",
    "    def get_filtered_df(self):\n",
    "        return self.dataframe\n",
    "\n",
    "    def extract_unique_uniprot(self):\n",
    "        unique_uniprot_values = pd.concat([self.dataframe['UniprotA'], self.dataframe['UniprotB']]).dropna().unique()\n",
    "        return pd.DataFrame({'isoform': unique_uniprot_values})\n",
    "\n",
    "    def get_filtered_df(self):\n",
    "        return self.dataframe\n",
    "\n",
    "class fill_list:\n",
    "    def __init__(self, dataset, protein_list, dataset_column, protein_list_column):\n",
    "        self.dataset = dataset\n",
    "        self.protein_list = protein_list\n",
    "        self.dataset_column = dataset_column\n",
    "        self.protein_list_column = protein_list_column\n",
    "        self.filter()\n",
    "\n",
    "    def filter(self):\n",
    "        dataset_column_1 = \"canonical_protein\"\n",
    "        dataset_column_2 = \"isoform\"\n",
    "        dataset_column_3 = \"gene_name\"\n",
    "        self.filtered_df = self.dataset[self.dataset[self.dataset_column].isin(self.protein_list[self.protein_list_column].dropna())][[dataset_column_1, dataset_column_2, dataset_column_3]]\n",
    "\n",
    "    def check_duplicates_protein_match(self): \n",
    "        duplicates_exist = self.filtered_df.duplicated(subset=[\"canonical_protein\"]).any()\n",
    "        if duplicates_exist:\n",
    "            print(\"DETECTED duplicates in 'canonical_protein'.\")\n",
    "        else:\n",
    "            print(\"There're NO duplicates in 'canonical_protein'.\")\n",
    "        return duplicates_exist\n",
    "\n",
    "    def get_filtered_df(self):\n",
    "        return self.filtered_df\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### LOAD FILES"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "iso_lib = fasta_to_df(\"UP000005640_9606_additional.fasta\")\n",
    "data_hct116 = \"Unique_interactions_HCT116.csv\"\n",
    "data_293t = \"Unique_interactions_293T.csv\"\n",
    "data_shared = \"Shared_interactions.csv\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### MAP TO ISOFORM LIBRARY"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Unique interactions HCT-116\n",
    "iso_uppi_hct116 = IsoRetriever(data_hct116).get_unique_isoforms_df() \n",
    "iso_lib_match_hct116 = iso_lib_matcher(iso_lib, iso_uppi_hct116).get_filtered_df()\n",
    "can_lib_match_hct116 = can_lib_matcher(iso_uppi_hct116, iso_lib).get_filtered_df()\n",
    "gene_lib_match_hct116 = gene_lib_matcher(iso_uppi_hct116, iso_lib).get_filtered_df()\n",
    "lib_match_hct116 = addition(iso_lib_match_hct116, can_lib_match_hct116, gene_lib_match_hct116, \"isoform\").get_merged_df()\n",
    "undefined_hct116 = subtract(iso_uppi_hct116, lib_match_hct116, \"isoform\", \"isoform\").subtraction()\n",
    "id_hct116 = SortUndefined(undefined_hct116).sort()\n",
    "index_hct116 = addition(lib_match_hct116, id_hct116, None, \"isoform\").get_merged_df()\n",
    "original_index_hct116 = index_hct116.copy()\n",
    "proteins_hct116 = IsoformCount(original_index_hct116).get_dataset_count()\n",
    "\n",
    "#Unique interactions 293T\n",
    "iso_uppi_293t = IsoRetriever(data_293t).get_unique_isoforms_df()\n",
    "iso_lib_match_293t = iso_lib_matcher(iso_lib, iso_uppi_293t).get_filtered_df()\n",
    "can_lib_match_293t = can_lib_matcher(iso_uppi_293t, iso_lib).get_filtered_df()\n",
    "gene_lib_match_293t = gene_lib_matcher(iso_uppi_293t, iso_lib).get_filtered_df()\n",
    "lib_match_293t = addition(iso_lib_match_293t, can_lib_match_293t, gene_lib_match_293t, \"isoform\").get_merged_df()\n",
    "undefined_293t = subtract(iso_uppi_293t, lib_match_293t, \"isoform\", \"isoform\").subtraction()\n",
    "id_293t = SortUndefined(undefined_293t).sort()\n",
    "index_293t = addition(lib_match_293t, id_293t, None, \"isoform\").get_merged_df()\n",
    "original_index_293t = index_293t.copy()\n",
    "proteins_293t = IsoformCount(original_index_293t).get_dataset_count()\n",
    "\n",
    "#Shared interactions\n",
    "iso_sppi = IsoRetriever(data_shared).get_unique_isoforms_df() \n",
    "iso_lib_match_shared = iso_lib_matcher(iso_lib, iso_sppi).get_filtered_df()\n",
    "can_lib_match_shared = can_lib_matcher(iso_sppi, iso_lib).get_filtered_df()\n",
    "gene_lib_match_shared = gene_lib_matcher(iso_sppi, iso_lib).get_filtered_df()\n",
    "lib_match_shared = addition(iso_lib_match_shared, can_lib_match_shared, gene_lib_match_shared, \"isoform\").get_merged_df()\n",
    "undefined_shared = subtract(iso_sppi, lib_match_shared, \"isoform\", \"isoform\").subtraction()\n",
    "id_shared = SortUndefined(undefined_shared).sort()\n",
    "index_shared = addition(lib_match_shared, id_shared, None, \"isoform\").get_merged_df()\n",
    "proteins_shared = IsoformCount(index_shared).get_dataset_count()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### ISOFORM DISTRIBUTION"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Unique interactions HCT-116\n",
    "print(\"HCT-116: Unique Interactions\")\n",
    "IsoformDistribution(proteins_hct116).plot_distribution()\n",
    "\n",
    "#Unique interactions 293T\n",
    "print(\"293T: Unique Interactions\")\n",
    "IsoformDistribution(proteins_293t).plot_distribution()\n",
    "\n",
    "#Shared interactions\n",
    "print(\"Shared Interactions\")\n",
    "IsoformDistribution(proteins_shared).plot_distribution()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### UNIQUE INTERACTIONS COMBINED"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Unique interactions HCT-116\n",
    "protein_match_hct116 = protein_matcher(proteins_hct116, proteins_293t, \"canonical_protein\", \"canonical_protein\").get_filtered_df()\n",
    "iso_ext_hct116 = IsoformExtractor(protein_match_hct116).get_unique_isoforms_df()\n",
    "handler_hct116 = UniqueInteractionsHandler(pd.read_csv(data_hct116))\n",
    "iso_hand_hct116 = handler_hct116.process_ppi(iso_ext_hct116)\n",
    "fill_red_ppi_hct116 = fill_list(index_hct116, iso_hand_hct116, \"isoform\", \"isoform\").get_filtered_df()\n",
    "\n",
    "\n",
    "#Unique interactions 293T\n",
    "protein_match_293t = protein_matcher(proteins_293t, proteins_hct116, \"canonical_protein\", \"canonical_protein\").get_filtered_df()\n",
    "iso_ext_293t = IsoformExtractor(protein_match_293t).get_unique_isoforms_df()\n",
    "handler_293t = UniqueInteractionsHandler(pd.read_csv(data_293t))\n",
    "iso_hand_293t = handler_293t.process_ppi(iso_ext_293t)\n",
    "fill_red_ppi_293t = fill_list(index_293t, iso_hand_293t, \"isoform\", \"isoform\").get_filtered_df()\n",
    "\n",
    "#Combined\n",
    "index_uppi = addition(fill_red_ppi_hct116, fill_red_ppi_293t, None, \"isoform\").get_merged_df()\n",
    "proteins_uppi = IsoformCount(index_uppi).get_dataset_count()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### ISOFORM DISTRIBUTION UNIQUE INTERACTIONS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"Combined Unique Interactions\")\n",
    "IsoformDistribution(proteins_uppi).plot_distribution()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "kex",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
